src/lua/ext_string.lua

```diff
@@ -0,0 +1,268 @@
+    if delta < 0 then
+        return inp
+    end
+    return char:rep(delta) .. inp
+end
+
+--- Center string in a field of given width.
+-- Prepend and append "(width - len(inp))/2" chars to given string.
+-- Input is never trucated.
+-- @function center
+-- @string       inp    the string
+-- @int          width  at least bytes to be returned
+-- @string[opt]  char   char of length 1 to fill with (" " by default)
+-- @returns             result string
+local function string_center(inp, width, char)
+    if type(inp) ~= 'string' then
+        error("string.center argument #1 should be a string", 2)
+    end
+    if type(width) ~= 'number' or width < 0 then
+        error("string.center argument #2 should be a positive integer", 2)
+    end
+    if char ~= nil and (type(char) ~= 'string' or #char ~= 1) then
+        error("string.center argument #3 should be a char (length 1) or nil", 2)
+    end
+    char = char or " "
+    local delta = width - #inp
+    if delta < 0 then
+        return inp
+    end
+    local pad_left = math.floor(delta / 2)
+    local pad_right = delta - pad_left
+    return char:rep(pad_left) .. inp .. char:rep(pad_right)
+end
+
+-- For now the best way to check, that string starts with sequence
+-- (with patterns disabled) is to cut line and check strings for equality
+
+--- Check that string (or substring) starts with given string
+-- Optionally restricting the matching with the given offsets
+-- @function startswith
+-- @string    inp     original string
+-- @string    head    the substring to check against
+-- @int[opt]  _start  start index of matching boundary
+-- @int[opt]  _end    end index of matching boundary
+-- @returns           boolean
+local function string_startswith(inp, head, _start, _end)
+    local head_len, inp_len = #head, #inp
+    if type(inp) ~= 'string' then
+        error("string.startswith argument #1 should be a string", 2)
+    end
+    if type(head) ~= 'string' then
+        error("string.startswith argument #2 should be a string", 2)
+    end
+    if _start ~= nil and type(_start) ~= 'number' then
+        error("string.startswith argument #3 should be an integer", 2)
+    end
+    if _end ~= nil and type(_end) ~= 'number' then
+        error("string.startswith argument #4 should be an integer", 2)
+    end
+    -- prepare input arguments (move negative values [offset from the end] to
+    -- positive ones and/or assign default values)
+    if _start == nil then
+        _start = 1
+    elseif _start < 0 then
+        _start = inp_len + _start + 1
+        if _start < 0 then _start = 0 end
+    end
+    if _end == nil or _end > inp_len then
+        _end = inp_len
+    elseif _end < 0 then
+        _end = inp_len + _end + 1
+        if _end < 0 then _end = 0 end
+    end
+    -- check for degenerate case (interval lesser than input)
+    if _end - _start + 1 < head_len then
+        return false
+    end
+    _end = _start + head_len - 1
+    return (inp:sub(_start, _end) == head)
```

Comment:
Please use ffi.C.memmem() instead. str:sub() allocates a new string and do one extra memcpy().